src/hotspot/share/memory/heap.cpp

```diff
@@ -303,7 +303,7 @@ void* CodeHeap::allocate(size_t instance_size) {
     mark_segmap_as_used(_next_segment, _next_segment + number_of_segments, false);
     block = block_at(_next_segment);
     block->initialize(number_of_segments);
-    _next_segment += number_of_segments;
+    Atomic::release_store(&_next_segment, _next_segment + number_of_segments);
```

Comment:
When we do the release here, what previous stores are we trying to order with the reading code below? It is hard to see the complete picture with this code just from the PR changes.

Reply:
@dholmes-ora you’re right, it is quite difficult to get the whole picture: let me try to make it a bit clearer (I hope it makes sense).

The issue originates here:
https://github.com/openjdk/jdk/blob/138542de7889e8002df0e15a79e31d824c6a0473/src/hotspot/share/services/memoryPool.cpp#L181-L182
We want to ensure that the values used to calculate `committed` are actually saved before the values used to calculate `used` are used.

The problem starts in  `CodeCache::allocate`. First `CodeHeap::allocate` is called and if there is not enough space we call `CodeHeap::expand_by` and retry `CodeHeap::allocate` (while loop):
https://github.com/openjdk/jdk/blob/e510dee162612d9a706ba54d0ab79a49139e77d8/src/hotspot/share/code/codeCache.cpp#L535-L537
During the second call of `CodeHeap::allocate` we update the values used to calculate the `used` variable (`_next_segment` in particular since it is the last to be updated) and we want to make sure that when we read it, all earlier values have been written.

As for the link between `used` and `_next_segment`, `used` is calculated here
https://github.com/openjdk/jdk/blob/e510dee162612d9a706ba54d0ab79a49139e77d8/src/hotspot/share/memory/heap.cpp#L556 from `_next_segment` and `_freelist_segments`. `_next_segment` is the last to be written and it is updated here
https://github.com/openjdk/jdk/blob/e510dee162612d9a706ba54d0ab79a49139e77d8/src/hotspot/share/memory/heap.cpp#L306 which is called from `CodeCache::allocate`
https://github.com/openjdk/jdk/blob/e510dee162612d9a706ba54d0ab79a49139e77d8/src/hotspot/share/code/codeCache.cpp#L535

Reply:
Sorry I'm still missing part of this (and I know we discussed it a few weeks ago). What is the variable(s) that is written before `_next_segment` that we need to see updated when we call `_codeHeap->capacity()`?

Reply:
Basically it would be the field `VirtualSpace::_high`. `_codeHeap->capacity()` calls `VirtualSpace::committed_size` which reads the `_high` and `_low` fields: https://github.com/openjdk/jdk/blob/e510dee162612d9a706ba54d0ab79a49139e77d8/src/hotspot/share/memory/virtualspace.cpp#L768-L770
`_low` doesn't really get updated but `_high` is increased in `VirtualSpace::expand_by`, which is called by `CodeHeap::expand_by` https://github.com/openjdk/jdk/blob/e510dee162612d9a706ba54d0ab79a49139e77d8/src/hotspot/share/memory/heap.cpp#L259 which in turn is called in `CodeCache::allocate`: https://github.com/openjdk/jdk/blob/e510dee162612d9a706ba54d0ab79a49139e77d8/src/hotspot/share/code/codeCache.cpp#L537

Reply:
@dholmes-ora is there possibly a way to restrict the update that we need to see to that `_high` variable only?

Reply:
So ... as we discussed offline a few weeks ago, it is very hard to clearly express the ordering constraints in this code as the two variables are so far apart in the code. So to reiterate, on the writer side we have:
```
 while (true) {
    cb = (CodeBlob*)heap->allocate(size);  // updates _next_segment
    if (cb != nullptr) break;
    if (!heap->expand_by(CodeCacheExpansionSize)) {  // updates _high
```
where the writes to `_high` in `expand_by` have to be ordered with the write to `_next_segment` in `allocate` that happen on the next loop iteration.

And on the reader side we have:
```
size_t used      = used_in_bytes();   // reads _next_segment
size_t committed = _codeHeap->capacity();  // reads _high
```
where the reads must also be ordered.

I can't help but think that the clearest fix here is to use explicit barriers as discussed before the locking solution was proposed. The problem using `release_store` and `load_acquire` in the bottom level code is that those semantics are only needed in this higher-level code for the `codeHeap`, not for every kind of `heap` the code applies to. 

I wonder if perhaps the following is clearer:
```
 while (true) {
    cb = (CodeBlob*)heap->allocate(size);
    if (cb != nullptr) break;
    if (!heap->expand_by(CodeCacheExpansionSize)) {
      // ...
    } else {
      OrderAccess::release(); // ensure heap expansion is visible before we allocate on next iteration
    }
```
coupled with:
```
size_t used      = used_in_bytes(); 
OrderAccess:acquire();
size_t committed = _codeHeap->capacity(); 
```

Thoughts?

Reply:
> I can't help but think that the clearest fix here is to use explicit barriers as discussed before the locking solution was proposed. The problem using `release_store` and `load_acquire` in the bottom level code is that those semantics are only needed in this higher-level code for the `codeHeap`, not for every kind of `heap` the code applies to.
> 
> I wonder if perhaps the following is clearer:
> 
> ```
>  while (true) {
>     cb = (CodeBlob*)heap->allocate(size);
>     if (cb != nullptr) break;
>     if (!heap->expand_by(CodeCacheExpansionSize)) {
>       // ...
>     } else {
>       OrderAccess::release(); // ensure heap expansion is visible

"visible to an asynchronous observer (e.g. CodeHeapPool::get_memory_usage())"

> before we allocate on next iteration
>     }
> ```
> 
> coupled with:
> 
> ```
> size_t used      = used_in_bytes(); 
> OrderAccess:acquire();
> size_t committed = _codeHeap->capacity(); 
> ```
> 
> Thoughts?

As long as it's more thoroughly commented, i guess so.

A floating release is, however, a pretty bad code smell. The only reason we're aren't solving this problem (IMHO) properly is that the code cache lock isn't re-entrant.

Reply:
>  The problem using `release_store` and `load_acquire` in the bottom level code is that those semantics are only needed in this higher-level code for the `codeHeap`, not for every kind of `heap` the code applies to.
> 
> I wonder if perhaps the following is clearer:
> 
> ```
>  while (true) {
>     cb = (CodeBlob*)heap->allocate(size);
>     if (cb != nullptr) break;
>     if (!heap->expand_by(CodeCacheExpansionSize)) {
>       // ...
>     } else {
>       OrderAccess::release(); // ensure heap expansion is visible before we allocate on next iteration
>     }
> ```

This would definitely make sense.

> A floating release is, however, a pretty bad code smell. The only reason we're aren't solving this problem (IMHO) properly is that the code cache lock isn't re-entrant.

Indeed, it still looks a bit "messy". I was also wondering if there could be a performance penalty by using explicit barriers (although at least `OrderAccess::release()` won't be called very often with this solution).

In any case I've modified the code to make use of explicit barriers following @dholmes-ora suggestion.

Thanks a lot!